File: go/parser.go
package main

import "fmt"

type Parser struct {
	lexer *Lexer
}

func NewParser(lexer *Lexer) *Parser {
	return &Parser{lexer: lexer}
}

func (p *Parser) Parse() *ASTNode {
	node, _ := p.parseCommandList()
	return node
}

func (p *Parser) parseCommandList() (*ASTNode, bool) {
	var nodes []*ASTNode
	for {
		tok := p.lexer.PeekToken()
		if tok.Type == TOK_EOF || tok.Type == TOK_RPAREN || tok.Type == TOK_RBRACE {
			break
		}
		// Reserved words that end command lists
		if tok.Value == "done" || tok.Value == "fi" || tok.Value == "esac" || tok.Value == "else" || tok.Value == "elif" {
			break
		}

		node := p.parseCommand()
		if node != nil {
			nodes = append(nodes, node)
		}

		tok = p.lexer.PeekToken()
		if tok.Type == TOK_SEMI || tok.Type == TOK_NEWLINE {
			p.lexer.NextToken()
		} else if tok.Type == TOK_AMP {
			p.lexer.NextToken()
			if node != nil {
				node.Conn = "&"
			}
		} else {
			// If not a separator, we might have reached the end of a line/command
			break
		}
	}

	if len(nodes) == 0 {
		return nil, false
	}
	if len(nodes) == 1 {
		return nodes[0], true
	}
	return &ASTNode{Type: "list", Pipes: nodes}, true
}

func (p *Parser) parseCommand() *ASTNode {
	tok := p.lexer.PeekToken()
	if tok.Type == TOK_EOF || tok.Type == TOK_NEWLINE || tok.Type == TOK_SEMI {
		return nil
	}

	switch tok.Value {
	case "if":
		return p.parseIf()
	case "while", "until":
		return p.parseWhile()
	case "for":
		return p.parseFor()
	case "case":
		return p.parseCase()
	case "{":
		p.lexer.NextToken()
		node, _ := p.parseCommandList()
		p.expect(TOK_RBRACE, "}")
		return &ASTNode{Type: "brace_group", Body: node}
	case "(":
		p.lexer.NextToken()
		node, _ := p.parseCommandList()
		p.expect(TOK_RPAREN, ")")
		return &ASTNode{Type: "subshell", Body: node}
	}

	return p.parseSimpleCommand()
}

func (p *Parser) parseSimpleCommand() *ASTNode {
	node := &ASTNode{Type: "simple", Args: []string{}}
	for {
		tok := p.lexer.PeekToken()
		if tok.Type == TOK_EOF || tok.Type == TOK_SEMI || tok.Type == TOK_NEWLINE || tok.Type == TOK_AMP || tok.Type == TOK_PIPE || tok.Type == TOK_AND_IF || tok.Type == TOK_OR_IF || tok.Type == TOK_RPAREN || tok.Type == TOK_RBRACE {
			break
		}

		fd := -1
		if tok.Type == TOK_IO_NUMBER {
			fdStr := p.lexer.NextToken().Value
			fmt.Sscanf(fdStr, "%d", &fd)
			tok = p.lexer.PeekToken()
		}

		if isRedirection(tok) {
			op := p.lexer.NextToken().Value
			if fd == -1 {
				if op == "<" || op == "<<" || op == "<<-" || op == "<&" || op == "<>" {
					fd = 0
				} else {
					fd = 1
				}
			}
			fileTok := p.lexer.NextToken()
			node.Redirs = append(node.Redirs, Redir{Op: op, File: fileTok.Value, Fd: fd})
			continue
		}

		if tok.Type == TOK_ASSIGNMENT && len(node.Args) == 0 {
			if node.Assigns == nil {
				node.Assigns = make(map[string]string)
			}
			parts := splitAssignment(tok.Value)
			node.Assigns[parts[0]] = parts[1]
			p.lexer.NextToken()
			continue
		}

		if tok.Type == TOK_WORD {
			node.Args = append(node.Args, tok.Value)
			p.lexer.NextToken()
			continue
		}

		break
	}

	// Handle pipeline and and/or logic
	tok := p.lexer.PeekToken()
	if tok.Type == TOK_PIPE {
		p.lexer.NextToken()
		right := p.parseCommand()
		return &ASTNode{Type: "pipeline", Pipes: []*ASTNode{node, right}}
	}
	if tok.Type == TOK_AND_IF {
		p.lexer.NextToken()
		right := p.parseCommand()
		return &ASTNode{Type: "and_or", Left: node, Right: right, Conn: "&&"}
	}
	if tok.Type == TOK_OR_IF {
		p.lexer.NextToken()
		right := p.parseCommand()
		return &ASTNode{Type: "and_or", Left: node, Right: right, Conn: "||"}
	}

	return node
}

func (p *Parser) parseIf() *ASTNode {
	p.lexer.NextToken() // if
	cond, _ := p.parseCommandList()
	p.expectValue("then")
	body, _ := p.parseCommandList()

	var elseNode *ASTNode
	tok := p.lexer.PeekToken()
	if tok.Value == "else" {
		p.lexer.NextToken()
		elseNode, _ = p.parseCommandList()
	} else if tok.Value == "elif" {
		elseNode = p.parseIf()
	}

	p.expectValue("fi")
	return &ASTNode{Type: "if", Cond: cond, Body: body, Else: elseNode}
}

func (p *Parser) parseWhile() *ASTNode {
	tok := p.lexer.NextToken()
	keyword := tok.Value
	cond, _ := p.parseCommandList()
	p.expectValue("do")
	body, _ := p.parseCommandList()
	p.expectValue("done")
	return &ASTNode{Type: keyword, Cond: cond, Body: body}
}

func (p *Parser) parseFor() *ASTNode {
	p.lexer.NextToken() // for
	varName := p.lexer.NextToken().Value
	p.expectValue("in")

	var words []string
	for {
		tok := p.lexer.PeekToken()
		if tok.Value == "do" || tok.Type == TOK_SEMI || tok.Type == TOK_NEWLINE {
			break
		}
		words = append(words, p.lexer.NextToken().Value)
	}

	if p.lexer.PeekToken().Type == TOK_SEMI {
		p.lexer.NextToken()
	}

	p.expectValue("do")
	body, _ := p.parseCommandList()
	p.expectValue("done")
	return &ASTNode{Type: "for", LoopVar: varName, LoopWords: words, Body: body}
}

func (p *Parser) parseCase() *ASTNode {
	p.lexer.NextToken() // case
	word := p.lexer.NextToken().Value
	p.expectValue("in")

	// Complex case parsing omitted for brevity in prototype parity
	for p.lexer.PeekToken().Value != "esac" && p.lexer.PeekToken().Type != TOK_EOF {
		p.lexer.NextToken()
	}
	p.expectValue("esac")
	return &ASTNode{Type: "case", Value: word}
}

func (p *Parser) expect(t TokenType, val string) {
	tok := p.lexer.NextToken()
	if tok.Type != t {
		// Panic for simplicity in prototype
	}
}

func (p *Parser) expectValue(val string) {
	tok := p.lexer.NextToken()
	if tok.Value != val {
		// Panic for simplicity
	}
}

func isRedirection(t Token) bool {
	return t.Type == TOK_LESS || t.Type == TOK_GREAT || t.Type == TOK_DLESS || t.Type == TOK_DGREAT || t.Type == TOK_LESSAND || t.Type == TOK_GREATAND || t.Type == TOK_LESSGREAT || t.Type == TOK_CLOBBER || t.Type == TOK_DLESSDASH
}

func splitAssignment(s string) []string {
	idx := 0
	for i, c := range s {
		if c == '=' {
			idx = i
			break
		}
	}
	return []string{s[:idx], s[idx+1:]}
}

File: src/parser.rs
use crate::lexer::Lexer;
use crate::types::{ASTNode, CaseItem, Redir, Token, TokenType};

pub struct Parser<'a> {
    lexer: &'a mut Lexer,
}

impl<'a> Parser<'a> {
    pub fn new(lexer: &'a mut Lexer) -> Self {
        Parser { lexer }
    }

    pub fn parse(&mut self) -> Option<ASTNode> {
        self.parse_command_list(true).0
    }

    fn parse_command_list(&mut self, top_level: bool) -> (Option<ASTNode>, bool) {
        let mut nodes = Vec::new();

        // Skip blank lines at the start of a top-level parse so they don't
        // make `parse()` return None and falsely terminate the script.
        if top_level {
            while self.peek_token().token_type == TokenType::Newline {
                self.lexer.next_token();
            }
        }

        loop {
            let tok = self.peek_token();
            if tok.token_type == TokenType::Eof
                || tok.token_type == TokenType::Rparen
                || tok.token_type == TokenType::Rbrace
            {
                break;
            }
            if Self::is_reserved_end(&tok.value) {
                break;
            }

            if let Some(node) = self.parse_command() {
                nodes.push(node);
            }

            let tok = self.peek_token();
            match tok.token_type {
                TokenType::Semi => {
                    self.lexer.next_token();
                }
                TokenType::Newline => {
                    self.lexer.next_token();
                    if top_level {
                        break;
                    }
                }
                TokenType::Ampersand => {
                    self.lexer.next_token();
                    if let Some(ref mut node) = nodes.last_mut() {
                        node.conn = "&".to_string();
                    }
                }
                _ => break,
            }
        }

        if nodes.is_empty() {
            return (None, false);
        }
        if nodes.len() == 1 {
            return (Some(nodes.remove(0)), true);
        }

        let pipes = nodes;
        let mut node = ASTNode::new("list");
        node.pipes = pipes;
        (Some(node), true)
    }

    fn is_reserved_end(value: &str) -> bool {
        // The lexer doesn't emit a dedicated Rbrace token for `}` — it
        // comes back as a Word — so we treat the literal value as a
        // command-list terminator alongside the other reserved words.
        matches!(
            value,
            "done" | "fi" | "esac" | "else" | "elif" | "then" | "do" | "}"
        )
    }

    fn peek_token(&mut self) -> Token {
        self.lexer.peek_token()
    }

    fn parse_command(&mut self) -> Option<ASTNode> {
        let tok = self.peek_token();

        if tok.token_type == TokenType::Eof
            || tok.token_type == TokenType::Newline
            || tok.token_type == TokenType::Semi
        {
            return None;
        }

        // POSIX function definition: name ( )  followed by compound command.
        if tok.token_type == TokenType::Word && self.is_posix_func_def() {
            return self.parse_posix_function();
        }

        if tok.value == "[[" {
            return Some(self.parse_dbracket());
        }

        match tok.value.as_str() {
            "if" => Some(self.parse_if()),
            "while" | "until" => Some(self.parse_while()),
            "for" => Some(self.parse_for()),
            "case" => Some(self.parse_case()),
            "{" => {
                self.lexer.next_token();
                let (body, _) = self.parse_command_list(false);
                self.expect_token_type(TokenType::Rbrace);
                let mut node = ASTNode::new("brace_group");
                node.body = body.map(Box::new);
                Some(node)
            }
            "(" => {
                self.lexer.next_token();
                let (body, _) = self.parse_command_list(false);
                self.expect_token_type(TokenType::Rparen);
                let mut node = ASTNode::new("subshell");
                node.body = body.map(Box::new);
                Some(node)
            }
            "!" => {
                self.lexer.next_token();
                let mut node = self.parse_command()?;
                node.bang = true;
                Some(node)
            }
            "function" => self.parse_function(),
            _ => self.parse_simple_command(),
        }
    }

    fn is_posix_func_def(&mut self) -> bool {
        let mut clone = self.lexer.clone();
        let name = clone.next_token();
        if name.token_type != TokenType::Word {
            return false;
        }
        // Reserved words and the "function" keyword don't form POSIX defs.
        if matches!(
            name.value.as_str(),
            "if" | "while"
                | "until"
                | "for"
                | "case"
                | "function"
                | "then"
                | "else"
                | "elif"
                | "fi"
                | "do"
                | "done"
                | "esac"
                | "in"
                | "{"
                | "}"
                | "!"
        ) {
            return false;
        }
        let lparen = clone.next_token();
        if lparen.token_type != TokenType::Lparen {
            return false;
        }
        let rparen = clone.next_token();
        rparen.token_type == TokenType::Rparen
    }

    fn parse_posix_function(&mut self) -> Option<ASTNode> {
        let name = self.lexer.next_token().value;
        self.lexer.next_token(); // (
        self.lexer.next_token(); // )
        self.skip_newlines();
        let body = self.parse_command();
        let mut node = ASTNode::new("function");
        node.func_name = name;
        node.func_body = body.map(Box::new);
        Some(node)
    }

    // [[ expr ]] conditional. The lexer hands us each inner token as-is
    // (including `&&`, `||`, `(`, `)`, and operator/operand words). We
    // collect raw values until `]]` and let the executor evaluate.
    fn parse_dbracket(&mut self) -> ASTNode {
        self.lexer.next_token(); // [[
        let mut args: Vec<String> = Vec::new();
        loop {
            let tok = self.lexer.next_token();
            if tok.token_type == TokenType::Eof {
                eprintln!("meowsh: syntax error: unexpected end of file in [[ ... ]]");
                break;
            }
            if tok.value == "]]" {
                break;
            }
            if tok.token_type == TokenType::Newline {
                continue;
            }
            args.push(tok.value);
        }
        let mut node = ASTNode::new("dbracket");
        node.args = args;
        node
    }

    fn parse_simple_command(&mut self) -> Option<ASTNode> {
        let mut node = ASTNode::new("simple");

        loop {
            let tok = self.peek_token();
            if Self::is_terminator(&tok) {
                break;
            }

            let mut fd = -1;
            if tok.token_type == TokenType::IoNumber {
                let fd_str = self.lexer.next_token().value;
                fd = fd_str.parse().unwrap_or(-1);
            }

            if Self::is_redirection(&tok) {
                let op = self.lexer.next_token().value;
                if fd == -1 {
                    if matches!(op.as_str(), "<" | "<<" | "<<-" | "<&" | "<>") {
                        fd = 0;
                    } else {
                        fd = 1;
                    }
                }
                let file_tok = self.lexer.next_token();
                let mut redir = Redir {
                    op: op.clone(),
                    file: file_tok.value.clone(),
                    fd,
                    heredoc_body: String::new(),
                };

                if op == "<<" || op == "<<-" {
                    redir.heredoc_body = self.lexer.read_heredoc(&file_tok.value, op == "<<-");
                }

                node.redirs.push(redir);
                continue;
            }

            if tok.token_type == TokenType::Assignment {
                if node.args.is_empty() {
                    let raw = tok.value.clone();
                    self.lexer.next_token();
                    let (name, value, is_append) = split_assignment(&raw);

                    // Array literal / append: `name=(...)` or `name+=(...)`.
                    if value.is_empty()
                        && self.peek_token().token_type == TokenType::Lparen
                    {
                        self.lexer.next_token(); // (
                        let mut items: Vec<String> = Vec::new();
                        loop {
                            let t = self.peek_token();
                            if t.token_type == TokenType::Rparen
                                || t.token_type == TokenType::Eof
                            {
                                break;
                            }
                            if t.token_type == TokenType::Newline
                                || t.token_type == TokenType::Semi
                            {
                                self.lexer.next_token();
                                continue;
                            }
                            items.push(self.lexer.next_token().value);
                        }
                        if self.peek_token().token_type == TokenType::Rparen {
                            self.lexer.next_token();
                        }
                        if is_append {
                            node.array_appends.insert(name, items);
                        } else {
                            node.array_assigns.insert(name, items);
                        }
                        continue;
                    }

                    if is_append {
                        node.scalar_appends.insert(name, value);
                    } else {
                        node.assigns.insert(name, value);
                    }
                    continue;
                } else {
                    node.args.push(tok.value.clone());
                    self.lexer.next_token();
                    continue;
                }
            }

            if tok.token_type == TokenType::Word {
                node.args.push(tok.value.clone());
                self.lexer.next_token();
                continue;
            }

            break;
        }

        let tok = self.peek_token();
        if tok.token_type == TokenType::Pipe {
            self.lexer.next_token();
            if let Some(right) = self.parse_command() {
                let left = node;
                let mut pipe_node = ASTNode::new("pipeline");
                pipe_node.pipes = vec![left, right];
                return Some(pipe_node);
            }
        }
        if tok.token_type == TokenType::AndIf {
            self.lexer.next_token();
            if let Some(right) = self.parse_command() {
                let mut and_or = ASTNode::new("and_or");
                and_or.left = Some(Box::new(node));
                and_or.right = Some(Box::new(right));
                and_or.conn = "&&".to_string();
                return Some(and_or);
            }
        }
        if tok.token_type == TokenType::OrIf {
            self.lexer.next_token();
            if let Some(right) = self.parse_command() {
                let mut and_or = ASTNode::new("and_or");
                and_or.left = Some(Box::new(node));
                and_or.right = Some(Box::new(right));
                and_or.conn = "||".to_string();
                return Some(and_or);
            }
        }

        Some(node)
    }

    fn is_terminator(tok: &Token) -> bool {
        matches!(
            tok.token_type,
            TokenType::Eof
                | TokenType::Semi
                | TokenType::Newline
                | TokenType::Ampersand
                | TokenType::Pipe
                | TokenType::AndIf
                | TokenType::OrIf
                | TokenType::Rparen
                | TokenType::Rbrace
        )
    }

    fn is_redirection(tok: &Token) -> bool {
        matches!(
            tok.token_type,
            TokenType::Less
                | TokenType::Great
                | TokenType::Dless
                | TokenType::Dgreat
                | TokenType::LessAnd
                | TokenType::GreatAnd
                | TokenType::LessGreat
                | TokenType::Clobber
                | TokenType::DlessDash
        )
    }

    fn parse_if(&mut self) -> ASTNode {
        self.lexer.next_token(); // if
        let (cond, _) = self.parse_command_list(false);

        self.expect_value("then");
        let (body, _) = self.parse_command_list(false);

        let else_node = {
            let tok = self.peek_token();
            match tok.value.as_str() {
                "else" => {
                    self.lexer.next_token();
                    self.parse_command_list(false).0.map(Box::new)
                }
                "elif" => Some(Box::new(self.parse_if())),
                _ => None,
            }
        };

        self.expect_value("fi");

        let mut node = ASTNode::new("if");
        node.cond = cond.map(Box::new);
        node.body = body.map(Box::new);
        node.else_body = else_node;
        node
    }

    fn parse_while(&mut self) -> ASTNode {
        let keyword = self.lexer.next_token().value;
        let (cond, _) = self.parse_command_list(false);
        self.expect_value("do");
        let (body, _) = self.parse_command_list(false);
        self.expect_value("done");

        let mut node = ASTNode::new(&keyword);
        node.cond = cond.map(Box::new);
        node.body = body.map(Box::new);
        node
    }

    fn parse_for(&mut self) -> ASTNode {
        self.lexer.next_token(); // for
        let var_name = self.lexer.next_token().value;

        self.expect_value("in");

        let mut words = Vec::new();
        loop {
            let tok = self.peek_token();
            if tok.value == "do"
                || tok.token_type == TokenType::Semi
                || tok.token_type == TokenType::Newline
            {
                break;
            }
            words.push(self.lexer.next_token().value);
        }

        if self.peek_token().token_type == TokenType::Semi {
            self.lexer.next_token();
        }

        self.expect_value("do");
        let (body, _) = self.parse_command_list(false);
        self.expect_value("done");

        let mut node = ASTNode::new("for");
        node.loop_var = var_name;
        node.loop_words = words;
        node.body = body.map(Box::new);
        node
    }

    fn parse_case(&mut self) -> ASTNode {
        self.lexer.next_token(); // case
        let word = self.lexer.next_token().value;

        self.skip_newlines();
        self.expect_value("in");
        self.skip_newlines();

        let mut node = ASTNode::new("case");
        node.value = word;
        node.cases = Vec::new();

        loop {
            let tok = self.peek_token();
            if tok.value == "esac" || tok.token_type == TokenType::Eof {
                break;
            }

            if tok.token_type == TokenType::Newline {
                self.lexer.next_token();
                continue;
            }

            if tok.value == "(" {
                self.lexer.next_token();
            }

            let pattern = self.lexer.next_token().value;
            self.expect_value(")");

            let (body, _) = self.parse_command_list(false);
            node.cases.push(CaseItem {
                pattern,
                body: body.map(Box::new),
            });

            let tok = self.peek_token();
            if tok.value == ";;" {
                self.lexer.next_token();
            }
            self.skip_newlines();
        }

        self.expect_value("esac");
        node
    }

    fn parse_function(&mut self) -> Option<ASTNode> {
        self.lexer.next_token(); // function
        let name = self.lexer.next_token().value;

        let body = self.parse_command();

        let mut node = ASTNode::new("function");
        node.func_name = name;
        node.func_body = body.map(Box::new);
        Some(node)
    }

    fn skip_newlines(&mut self) {
        loop {
            let tok = self.peek_token();
            if tok.token_type != TokenType::Newline {
                break;
            }
            self.lexer.next_token();
        }
    }

    fn expect_token_type(&mut self, _token_type: TokenType) {
        self.lexer.next_token();
    }

    fn expect_value(&mut self, val: &str) {
        let tok = self.lexer.next_token();
        if tok.value != val {
            // Error handling
        }
    }
}

pub fn parse(line: &str) -> Option<ASTNode> {
    let mut lexer = Lexer::new(line);
    let mut parser = Parser::new(&mut lexer);
    parser.parse()
}

// Splits an Assignment-token value `name=val` or `name+=val` into
// (name, val, is_append).
fn split_assignment(raw: &str) -> (String, String, bool) {
    if let Some(idx) = raw.find("+=") {
        return (raw[..idx].to_string(), raw[idx + 2..].to_string(), true);
    }
    if let Some(idx) = raw.find('=') {
        return (raw[..idx].to_string(), raw[idx + 1..].to_string(), false);
    }
    (raw.to_string(), String::new(), false)
}
